confluent_kafka/consumer.py

```diff
@@ -0,0 +1,145 @@
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
+        if not on_commit:
+            on_commit = conf.get('on_commit', None)
+
+        if not stats_cb:
+            stats_cb = conf.get('stats_cb', None)
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+        super(Consumer, self).__init__(conf, on_commit=on_commit, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def poll(self, timeout=-1.0, key_deserializer=None, value_deserializer=None):
+        """
+            Consumes a message, triggers callbacks, returns an event.
+
+            The application must check the returned Message object’s Message.error() method to distinguish
+            between proper messages, an error(see error().code() for specifics), or an event.
+
+            :param float timeout:  Maximum time in seconds to block waiting for message, event or callback.
+            :param func key_deserializer(topic, key): Converts message key bytes to object.
+                **note** deserializers are responsible for handling NULL keys
+            :param func value_deserializer(topic, value): Converts message value bytes to object.
+                **note** deserializers are responsible for handling NULL values
+            :returns: A confluent_kafka.Message or None on timeout.
+            :raises RuntimeError: If called on a closed consumer.
+        """
+
+        msg = super(Consumer, self).poll(timeout)
+
+        if not msg or msg.error():
+            return msg
+
+        topic = msg.topic()
+
+        # parameter overrides take precedence over instance functions
+        if not key_deserializer:
+            key_deserializer = self._key_deserializer
+
+        if key_deserializer:
+            msg.set_key(key_deserializer(topic, msg.key()))
+
+        if not value_deserializer:
+            value_deserializer = self._value_deserializer
+
+        if value_deserializer:
+            msg.set_value(value_deserializer(topic, msg.value()))
+
+        return msg
+
+    def consume(self, num_messages=1, timeout=-1):
```

Comment:
1. I'd question the need for both consume and poll at the python api level. also, the justification for the naming is going to be non-obvious to many people and cause friction. batch consume is old skool librdkafka right? can we just get rid of / depreciate `consume`?
2. if it's not depreciated, I think we probably want to enable deserializers. that you don't take this method seriously (barf out if the serializers are not there) is an indication it should be depreciated...
3. batch consuming could actually be valuable (even in .NET I think it would represent a significant perf improvement), lets discuss.

Reply:
1. I would like to deprecate batch consuming with the python client entirely. 
2. see 1 
3. In order to do batch consumption correctly we would want to error out if `auto.offset.reset=true`. This would need to be accompanied with an example or two of how you would appropriately manage offset commits when using the batch consumer.

Reply:
auto.offset.reset=true -> enable.auto.offset.store=true

Reply:
I agree with Matt on 2, consume() should support serdes even with num_messages > 1.
The offset-store problem does not have anything to do with serdes being used or not.

Reply:
The `offset-store problem` was not specific to serialziation but instead to the use of Consume, and thus batch consuming, in general. 

I actually decided to forego implementing deserialization in the batch consumer based on a conversation you and I had previously. I'll go ahead and add it though as it seems there has been a change of heart. 

The question remains however. Do we or do we not want to deprecate the use of batch consumption.